### PR TITLE
Added support and tests on RHEL systems

### LINKED ISSUE #14: remove cloud_platform variable
should no longer be needed after switch to geerlingguy's roles

### LINKED ISSUE #16: fix role on CentOS and Fedora
in PR #15 I've enabled travis builds on centos and fedora.

The travis build fails immediately, since they don't have `/usr/bin/python3`.
https://travis-ci.org/marvel-nccr/ansible-role-aiida/jobs/636646808#L494

@espenfl Can I leave this for you?

### PR DESCRIPTION
**Pull request by espenfl**
fixes #14 
fixes #16 

Added support and tests on RHEL systems

### CODE REVIEW

File: meta/main.yml [hunk 1]
```diff
@@ -19,3 +19,8 @@ galaxy_info:
 dependencies:
   - role: marvel-nccr.current_user
     version: master
+  - name: ansible-role-rabbitmq
+    src: https://github.com/claudineimatos/ansible-role-rabbitmq.git
+    scm: git
+    version: master
+    when: ansible_distribution == "CentOS"
```
Comment (ltalirz) [inline meta/main.yml]: so, what is the policy here, i.e.:
is the idea not to depend on external roles for rabbitmq or postgres - then I guess we can simply hardcode things inside our role
or is the idea to use rabbitmq/postgres roles that work across OSs - then I guess we should  use it for all OSs, and not just for Centos, right?

File: molecule/default/molecule.yml [hunk 2]
```diff
@@ -10,7 +10,8 @@ lint:
     config-file: molecule/default/yamllint.yml
 platforms:
   - name: instance
-    image: "geerlingguy/docker-${MOLECULE_DISTRO:-ubuntu1804}-ansible:latest"
+    # image: "geerlingguy/docker-${MOLECULE_DISTRO:-ubuntu1804}-ansible:latest"
+    image: "geerlingguy/docker-centos8-ansible:latest"
```
Comment (ltalirz) [inline molecule/default/molecule.yml]: why this change?

File: molecule/default/requirements.yml [hunk 3]
```diff
@@ -1,3 +1,7 @@
 ---
 - src: marvel-nccr.current_user
   version: master
+#- name: ansible-role-rabbitmq
```
Comment (ltalirz) [inline molecule/default/requirements.yml]: does this not work?

**Review by ltalirz — COMMENTED** [review]
just some questions but I guess this is anyhow work in progress
I'll look at the travis logs as well now

File: tasks/aiida-core.yml [hunk 4]
```diff
@@ -16,11 +16,14 @@
...
     name:
-      - "${HOME}/src/aiida-core"
+      - "{{ aiida_source_folder }}/aiida-core"
     virtualenv: "{{ aiida_venv }}"
     editable: true
     # extra_args: --no-use-pep517
   register: aiida_core_install
+  # make sure we update the LC_ALL (we changed the file, but somehow it does not update)
+  environment:
+    LC_ALL: "C.UTF-8"
```
Comment (ltalirz) [inline tasks/aiida-core.yml]: Is this still needed?

File: tasks/aiida-daemon.yml [hunk 5]
```diff
@@ -18,6 +18,14 @@
...
     daemon_user_home: "{{ current_user_home }}"
     daemon_aiida_venv: "{{ aiida_venv |  regex_replace('\\$\\{HOME}', current_user_home) }}"
+  when: ansible_service_mgr == "systemd"
+
+- name: Check if the AiiDA daemon is running
+  shell: pgrep -a verdi | grep run_daemon | awk '{print $1}'
+  changed_when: false
+  register: aiida_daemon_status
+  failed_when: aiida_daemon_status.rc not in [0,1]
+  when: ansible_service_mgr != "systemd"
```
Comment (ltalirz) [inline tasks/aiida-daemon.yml]: How about looking for `not running` in the output of `verdi daemon status`?

File: tasks/aiida-daemon.yml [hunk 6]
```diff
@@ -28,4 +36,14 @@
...
+  when: ansible_service_mgr == "systemd"
+
+- name: Start AiiDA Daemon
+  become: true
+  become_user: "{{ root_user }}"
+  command: "{{ daemon_aiida_venv }}/bin/verdi -p {{ aiida_profile_name }} daemon start"
+  vars:
+    daemon_user: "{{ current_user }}"
+    daemon_user_home: "{{ current_user_home }}"
+    daemon_aiida_venv: "{{ aiida_venv |  regex_replace('\\$\\{HOME}', current_user_home) }}"
```
Comment (ltalirz) [inline tasks/aiida-daemon.yml]: how are these variables used here?

File: tasks/aiida-deps-debian.yml [hunk 7]
```diff
@@ -1,10 +1,15 @@
 ---
+- name: Update apt cache
+  become: true
+  become_user: "{{ root_user }}"
+  apt:
+    update_cache: true
```
Comment (ltalirz) [inline tasks/aiida-deps-debian.yml:6]: does one need to set a time to live so that this does not cause changes?

File: tasks/aiida-prepare.yml [hunk 8]
```diff
@@ -1,19 +1,47 @@
...
+  become_user: "{{ aiida_postgres_user }}"
+  command: "{{ aiida_postgres_bin_location }}/pg_ctl start -D {{ aiida_postgres_data_location }} -s -o '-p 5432' -w -t 300"
+  when: postgres_status.rc !=0 and ansible_service_mgr != "systemd"
+
+- name: Check if RabbitMQ is running
+  shell: pgrep rabbitmq
+  changed_when: false
+  register: rabbitmq_status
+  failed_when: rabbitmq_status.rc not in [0,1]
+  when: ansible_service_mgr != "systemd"
```
Comment (ltalirz) [inline tasks/aiida-prepare.yml]: perhaps move the non-systemd stuff into a 
```yaml
- block:
   - name: Check if ...
   - name: Start p
   when: ansible_service_mgr != "systemd"
```

**Review by ltalirz — COMMENTED** [review]
some comments looking through...

File: tasks/aiida-daemon.yml (hunk 5 above)
Comment (espenfl) [reply tasks/aiida-daemon.yml]: That we could do, but would have to change it. Now it is fairly general unless we change the name of the process. What do you think?

File: tasks/aiida-core.yml (hunk 4 above)
Comment (espenfl) [reply tasks/aiida-core.yml]: Nope, do not think so. Removed.

File: tasks/aiida-daemon.yml (hunk 6 above)
Comment (espenfl) [reply tasks/aiida-daemon.yml]: Cleaned. In fact only the last one is needed.

File: tasks/aiida-deps-debian.yml (hunk 7 above)
Comment (espenfl) [reply tasks/aiida-deps-debian.yml:6]: That I do not know. Not very familiar with `apt`. What should I do?

File: tasks/aiida-prepare.yml (hunk 8 above)
Comment (espenfl) [reply tasks/aiida-prepare.yml]: That is certainly a good suggestion. Done.

File: tasks/aiida-daemon.yml [hunk 9]
```diff
@@ -28,4 +29,19 @@
...
+
+- block:
+    - name: Check if the AiiDA daemon is running
+      shell: pgrep -a verdi | grep run_daemon | awk '{print $1}'
+      changed_when: false
+      register: aiida_daemon_status
+      failed_when: aiida_daemon_status.rc not in [0,1]
+    - name: Start AiiDA Daemon
+      become: true
+      become_user: "{{ root_user }}"
```
Comment (ltalirz) [inline tasks/aiida-daemon.yml]: If you just start the daemon (not a system service), I don't think you need to be the root user.

File: tasks/aiida-daemon.yml [hunk 10]
```diff
@@ -28,4 +29,19 @@
...
+      command: "{{ daemon_aiida_venv }}/bin/verdi -p {{ aiida_profile_name }} daemon start"
+      vars:
+        daemon_aiida_venv: "{{ aiida_venv |  regex_replace('\\$\\{HOME}', current_user_home) }}"
```
Comment (ltalirz) [inline tasks/aiida-daemon.yml]: this can then also be simplified

File: tasks/aiida-deps-debian.yml (hunk 7 above)
Comment (ltalirz) [reply tasks/aiida-deps-debian.yml:6]: e.g. set `cache_valid_time: 3600` for 1h
I'm surprised this does not trigger the idempotency check, since the default value [is zero](https://docs.ansible.com/ansible/latest/modules/apt_module.html)

Maybe it checks the result and would fail only if there actually were some updates?

File: tasks/aiida-deps-debian.yml [hunk 11]
```diff
@@ -20,5 +25,30 @@
       - python3-psycopg2
       - python3-gi
       - python3-gi-cairo
+      - python3-venv
       - gir1.2-gtk-3.0
       - python3-tk
+
+- name: Fetch location of the Postgresql binaries
+  command: find /usr/lib/postgresql/ -name bin
```
Comment (ltalirz) [inline tasks/aiida-deps-debian.yml]: Just for context, [pgtest](https://github.com/jamesnunn/pgtest/blob/master/pgtest/pgtest.py#L100) in the end does the same, so I guess there really is no better way.

File: tasks/aiida-deps-debian.yml [hunk 12]
```diff
@@ -20,5 +25,30 @@
...
+  changed_when: false
+  register: aiida_postgres_bin_raw
+  when: ansible_service_mgr != "systemd"
+
+- name: Set location of the Postgresql binaries
+  set_fact:
+    aiida_postgres_bin_location: "{{ aiida_postgres_bin_raw.stdout }}"
```
Comment (ltalirz) [inline tasks/aiida-deps-debian.yml]: This is rather easy to break - e.g. I have systems where I have multiple versions of postgresql installed.
How about this:
```suggestion
    aiida_postgres_bin_location: "{{ aiida_postgres_bin_raw.stdout_lines[0] }}"
```

File: tasks/aiida-deps-debian.yml [hunk 13]
```diff
@@ -20,5 +25,30 @@
...
+  when: ansible_service_mgr != "systemd"
+
+- name: Set data location for Postgresql
+  set_fact:
+    aiida_postgres_data_location: /var/lib/postgresql/data
+  when: ansible_service_mgr != "systemd"
```
Comment (ltalirz) [inline tasks/aiida-deps-debian.yml]: Again this whole thing is perhaps best put into a block

File: tasks/aiida-deps-rhel.yml [hunk 14]
```diff
@@ -2,47 +2,115 @@
...
+
+- name: Initialize database
+  become: true
+  become_user: "{{ aiida_postgres_user }}"
+  command: "{{ aiida_postgres_bin_location }}/initdb {{ aiida_postgres_data_location }}"
+  args:
+    creates: "{{ aiida_postgres_data_location }}/pg_hba.conf"
+  when: ansible_service_mgr != "systemd"
+
+# Avoid password on postgres, fix in future so that we can use md5
```
Comment (ltalirz) [inline tasks/aiida-deps-rhel.yml:113]: please open an issue for this

File: tasks/aiida-daemon.yml [hunk 15]
```diff
@@ -28,4 +29,19 @@
     masked: false
     daemon-reload: true
     state: restarted
-  when: not (cloud_platform is defined and cloud_platform == 'docker')
+  when: ansible_service_mgr == "systemd"
+
+- block:
+    - name: Check if the AiiDA daemon is running
```
Comment (ltalirz) [inline tasks/aiida-daemon.yml:36]: Please add a comment to replace this with `verdi daemon status` after aiida 1.1.0
Added PR to fix this https://github.com/aiidateam/aiida-core/pull/3729

File: tasks/aiida-prepare.yml [hunk 16]
```diff
@@ -37,11 +66,24 @@
...
+  pip:
+    name:
+      - virtualenvwrapper
+
+- name: Install pip packages in the AiiDA virtual environment
   pip:
     name:
       - pip~=19.3
       - numpy==1.16.4
+      # Flask is required for the AiiDA REST API
```
Comment (ltalirz) [inline tasks/aiida-prepare.yml]: That should be taken care of by installing the `rest` extra of the AiIDA package

File: tasks/aiida-rest.yml [hunk 17]
```diff
@@ -16,13 +16,30 @@
...
   systemd:
     name: aiida-rest@{{ aiida_profile_name }}
     enabled: "{{ aiida_rest_service_enabled }}"
     masked: false
     daemon-reload: true
-  when: not (cloud_platform is defined and cloud_platform == 'docker')
+  when: ansible_service_mgr == "systemd"
+
+- block:
+    - name: Check if the AiiDA REST API is running
```
Comment (ltalirz) [inline tasks/aiida-rest.yml:32]: Perhaps better check that it is listening on the port we expect, i.e. [something like](https://stackoverflow.com/questions/28606876/ansible-check-if-service-is-listening-on-a-specific-port)

```
command: curl -sf localhost:5000/api/v4/server
```

File: tasks/aiida-rest.yml [hunk 18]
```diff
@@ -16,13 +16,30 @@
...
+      shell: pgrep -a verdi | grep restapi | awk '{print $1}'
+      changed_when: false
+      register: aiida_rest_status
+      failed_when: aiida_rest_status.rc not in [0,1]
+
+    - name: Start AiiDA REST API
+      become: true
```
Comment (ltalirz) [inline tasks/aiida-rest.yml]: again, the user can do this

**Review by ltalirz — CHANGES_REQUESTED** [review]
thanks a lot @espenfl  - I think this generalization of the role is very useful and I hope it will help admins installing AiiDA going forward - we'll just need to get people to use it!

Most of my requests are minor.
There's of course the general issue that these `pgrep` checks for running processes are not resilient and should be replaced by alternatives where possible. Anyhow, we can open an issue for that and do it later.

File: tasks/aiida-daemon.yml (hunk 9 above)
Comment (espenfl) [reply tasks/aiida-daemon.yml]: Yes. Thanks for spotting the mistake. Fixed.

File: tasks/aiida-daemon.yml (hunk 10 above)
Comment (espenfl) [reply tasks/aiida-daemon.yml]: Fixed

File: tasks/aiida-deps-debian.yml (hunk 11 above)
Comment (espenfl) [reply tasks/aiida-deps-debian.yml]: I do not know. Possibly not.

File: tasks/aiida-deps-debian.yml (hunk 12 above)
Comment (espenfl) [reply tasks/aiida-deps-debian.yml]: Agreed. Simple and easy to explain. Fixed.

File: tasks/aiida-deps-rhel.yml (hunk 14 above)
Comment (espenfl) [reply tasks/aiida-deps-rhel.yml:113]: Will do. Fixed.

File: tasks/aiida-daemon.yml (hunk 15 above)
Comment (espenfl) [reply tasks/aiida-daemon.yml:36]: Fixed.

File: tasks/aiida-prepare.yml (hunk 16 above)
Comment (espenfl) [reply tasks/aiida-prepare.yml]: Somehow it did not. Removing now and then we will see how it goes.

File: tasks/aiida-rest.yml (hunk 17 above)
Comment (espenfl) [reply tasks/aiida-rest.yml:32]: This only checks that the system is running and if we should start it again. My suggestion is that if we are going to check that things are actually running correctly I think we should do that in a separate PR. This also includes running a few AiiDA tests etc. Agreed?

File: tasks/aiida-rest.yml (hunk 18 above)
Comment (espenfl) [reply tasks/aiida-rest.yml]: Indeed. Thanks for pointing this out. Fixed.

File: tasks/aiida-rest.yml (hunk 17 above)
Comment (ltalirz) [reply tasks/aiida-rest.yml:32]: Fine.

Comment (espenfl) [reply tasks/aiida-rest.yml:32]: Awesome.

File: tasks/aiida-deps-debian.yml (hunk 13 above)
Comment (espenfl) [reply tasks/aiida-deps-debian.yml]: Indeed indeed. Fixed.

**Comment by espenfl** [comment]
> e.g. set cache_valid_time: 3600 for 1h

Not sure why that was marked as resolved. Should we not just update the cache on every trigger of `ansible`? I find that maybe the safest. Maybe not?

>I'm surprised this does not trigger the idempotency check, since the default value is zero
>Maybe it checks the result and would fail only if there actually were some updates?

Yes I suspect so, but have not actually checked.

File: tasks/aiida-prepare.yml (hunk 16 above)
Comment (ltalirz) [reply tasks/aiida-prepare.yml]: looking at where the git repo is installed, it seems to me we're simply not installing the `rest` extra

File: tasks/aiida-rest.yml (hunk 17 above)
Comment (ltalirz) [reply tasks/aiida-rest.yml]: Just 'unresolving' this to remember the fix when we do it.

File: tasks/aiida-prepare.yml (hunk 16 above)
Comment (espenfl) [reply tasks/aiida-prepare.yml]: Aha, tests are running fine now so that was probably added when one of the plugins was enabled. Maybe. By the way do we want to have the plugin installation enabled by default?

Comment (ltalirz) [reply tasks/aiida-prepare.yml]: Ok, I see... I'm fine with disabling it by default. 

This aiida role really started as a role to install the aiida environment for the Quantum Mobile. Eventually I might move out the plugins/codes/examples to a separate role.

Comment (ltalirz) [reply tasks/aiida-prepare.yml]: As long as it is in this role, however, we should keep testing it (i.e. please keep it in the molecule playbook)

**Comment by ltalirz** [comment]
Thanks for your work @espenfl - let me know once you think it's ready to be merged.
I see no barriers to merging this.

**Comment by espenfl** [comment]
Happy to help. Let us merge.

**Review by ltalirz — APPROVED** [review]
let's go!

**Comment by ltalirz** [comment]
What did I say about the apt_cache ;-) 
https://travis-ci.org/marvel-nccr/ansible-role-aiida/jobs/641438706#L758

Anyhow, I'll fix this as well as the rest extra

**Comment by espenfl** [comment]
Great. Did not consider that ;) In fact this is an example of tweaking to pass a test which yields a behavior you not necessarily 
want in actual use (in case apt have some updates put in during the time slot that is). Of course the benefits as a whole of idempotence outweights this I guess. Thanks for fixing the remaining bits.

**Comment by ltalirz** [comment]
> In fact this is an example of tweaking to pass a test which yields a behavior you not necessarily want in actual use (in case apt have some updates put in during the time slot that is).

I'm not sure that is true. Note that even if new versions of a package appear in between ansible runs, our role will actually *not* update them (default of `apt:` is `state:present`, not `state:latest`).

If I understand the ansible philosophy correctly, then in the "strict interpretation" `apt-get update` should actually be executed only once and never again.
The way to handle apt package updates via ansible is through setting up an automatic cron job (we do this on Materials Cloud servers for security updates) - it's not the job of the provisioner.
Of course, when you change the role and require a different *minimal* version of a package, then this should result in a changed state.

**Comment by espenfl** [comment]
Good points.
